website/blog/2023-04-04-presto-hudi-deep-dive.md

```diff
@@ -0,0 +1,110 @@
+---
+author: Pratyaksh Sharma
+authorURL: https://www.linkedin.com/in/pratyaksh-sharma-60593769/
+title: Presto Hudi Deep Dive
+---
+
+With the growing popularity of the lakehouse approach, it has become increasingly important for query engines to support these new formats such as Hudi. A [previous blog](https://prestodb.io/blog/2020/08/04/prestodb-and-hudi) discusses the evolution of presto-hudi integration via hive connector at a high level. With the latest community developments, a [separate presto-hudi connector](https://github.com/prestodb/presto/issues/17006) has come up but it is not at par with the hive connector in terms of security features, caching and cost-based optimization. Hive connector has [CachingDirectoryLister](https://github.com/prestodb/presto/blob/master/presto-hive/src/main/java/com/facebook/presto/hive/CachingDirectoryLister.java) which can be used for caching the splits for a given table for a configurable period of time. It also supports [analyze command](https://prestodb.io/docs/current/sql/analyze.html) which helps in better planning during optimization phase. In this blog, we dive deeper into presto-hudi integration supported via hive connector from a developer's perspective and list down the various checkpoints where the query execution moves from presto to hudi library.
+
+<!--truncate-->
+
+## Input formats and Record readers
+Before actually diving into the presto-hudi integration, let us first review a few related concepts. In the Hadoop world, data is stored in a logical file. This file can have various physical layouts (format) on remote storage. A "Path" identifies a file. Further, an "input format" abstraction allows reading portions (splits) of a file. To keep it short and sweet, input formats help generate splits from files which are logical representations of the data. Input formats provide record readers which do the actual reading from these splits.
+
+InputFormat interface exposes [2 methods](https://hadoop.apache.org/docs/stable/api/org/apache/hadoop/mapred/InputFormat.html) - `getSplits()` and `getRecordReader()`.
+
+### Creation of splits
+`getSplits()` method facilitates the creation of splits from actual data files. On a high level, this involves reading the input paths from the configuration object and scanning those paths to get the files present. Files are logically represented as FileStatus. These file paths are then passed to the `makeSplit()` method which creates the splits for further processing. Below are the snippets of the actual code where these calls are being made.
+
+![](/img/blog/2023-04-04-presto-hudi-deep-dive/get-splits-call.png)
```

Comment:
Would it be better to share a permalink to the code instead of screenshots?

Reply:
I would prefer adding permalink along with having screenshot here. The reason is with code changes, permalink can start pointing to some other piece of code. Will do the needful.